File: packages/shared-ts/src/telemetry.ts
// W3C traceparent propagation (instructions.md §8.1, §19).
// Wire-compatible with packages/shared-py/olma_shared/telemetry.py.
import { randomBytes } from "node:crypto";

const TP_RE = /^([0-9a-f]{2})-([0-9a-f]{32})-([0-9a-f]{16})-([0-9a-f]{2})$/;

export interface SpanContext {
  traceId: string; // 32 hex
  spanId: string; // 16 hex
  sampled: boolean;
}

const hex = (n: number): string => randomBytes(n).toString("hex");

export const toTraceparent = (c: SpanContext): string =>
  `00-${c.traceId}-${c.spanId}-${c.sampled ? "01" : "00"}`;

export const newTrace = (sampled = true): SpanContext => ({
  traceId: hex(16),
  spanId: hex(8),
  sampled,
});

export function parse(traceparent: string): SpanContext | null {
  const m = TP_RE.exec((traceparent || "").trim());
  if (!m) return null;
  const [, , traceId, spanId, flags] = m;
  if (traceId === "0".repeat(32) || spanId === "0".repeat(16)) return null;
  return { traceId, spanId, sampled: (parseInt(flags, 16) & 0x01) === 0x01 };
}

export function child(traceparent: string | null | undefined): SpanContext {
  const parent = traceparent ? parse(traceparent) : null;
  if (!parent) return newTrace();
  return { traceId: parent.traceId, spanId: hex(8), sampled: parent.sampled };
}

File: packages/shared-py/olma_shared/telemetry.py
"""W3C traceparent propagation (instructions.md §8.1, §19).

Prod plugs the OpenTelemetry SDK (traces → Tempo); this module handles the wire
format — generate a root trace, parse an incoming header, derive a child span —
so every service can propagate `traceparent` today without the full SDK.

Format: 00-<32 hex trace_id>-<16 hex span_id>-<2 hex flags>
"""

from __future__ import annotations

import os
import re
from dataclasses import dataclass

_TP_RE = re.compile(r"^([0-9a-f]{2})-([0-9a-f]{32})-([0-9a-f]{16})-([0-9a-f]{2})$")


@dataclass(frozen=True)
class SpanContext:
    trace_id: str  # 32 hex
    span_id: str   # 16 hex
    sampled: bool = True

    def to_traceparent(self) -> str:
        flags = "01" if self.sampled else "00"
        return f"00-{self.trace_id}-{self.span_id}-{flags}"


def _rand_hex(n_bytes: int) -> str:
    return os.urandom(n_bytes).hex()


def new_trace(sampled: bool = True) -> SpanContext:
    """Start a fresh root trace."""
    return SpanContext(_rand_hex(16), _rand_hex(8), sampled)


def parse(traceparent: str) -> SpanContext | None:
    """Parse an incoming header; return None if malformed (caller starts a new trace)."""
    m = _TP_RE.match((traceparent or "").strip())
    if not m:
        return None
    _, trace_id, span_id, flags = m.groups()
    if trace_id == "0" * 32 or span_id == "0" * 16:
        return None  # invalid per spec
    return SpanContext(trace_id, span_id, bool(int(flags, 16) & 0x01))


def child(traceparent: str | None) -> SpanContext:
    """Continue an incoming trace with a new span, or start a new root if absent/invalid."""
    parent = parse(traceparent) if traceparent else None
    if parent is None:
        return new_trace()
    return SpanContext(parent.trace_id, _rand_hex(8), parent.sampled)
